src/main/ruby/core/array.rb

```diff
@@ -1311,6 +1311,27 @@ def recursively_flatten(array, out, max_levels = -1)
 
   private :recursively_flatten
 
+  def sort(&block)
+    sort_impl(block)
+  end
+
+  # Currently primitives do not support a block argument
+  private def sort_impl(block)
+    Truffle.primitive :array_sort
+
+    sorted = dup
+    Truffle.privately do
+      if block
+        sorted.mergesort_block!(block)
+      else
+        sorted.mergesort!
+      end
+    end
+
+    ret = []
+    Truffle::Array.steal_storage(ret, sorted)
```

Comment:
Out of scope here, but I think some of the logic got broken here along the way. IIRC, the reason to use `steal_storage` was so work could be done in a temporary buffer and then update the store in an existing array reference. I.e., to support `Array` methods that do in-place operations. But I can't see any reason why `sorted` couldn't just be used as the return value here.

Reply:
I wondered too. Maybe we need to return an instance of Array and not a subclass?

Reply:
Ahh, i think that's it.

Reply:
Well, now that it's not a snippet, we can add a comment ;-)